Change package name to badger_amcl

12tbuchman opened the pull request:
The name is specified in both package.xml and CMakeLists.txt.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)
<!-- Reviewable:end -->

12tbuchman committed:
Change package name and namespace to badger_amcl

The package name is specified in both package.xml and CMakeLists.txt and is needed
in the AMCL.cfg configuration file.

12tbuchman committed:
Remove unneded parameter from node.cpp

12tbuchman committed:
Fix bug in node_3d getFootprintToFrameTransform

Function is supposed to return a boolean but was not returning anything.

12tbuchman committed:
Remove wait_for_occupancy_map from octomap

This is no longer needed in the octommap and is managed by node_3d.

Improve whitespace in particle filter and point cloud scanner.

12tbuchman committed:
Fix bug in calcOdomDelta

12tbuchman committed:
Improve spacing in odom.cpp

12tbuchman committed:
Improve spacing in particle_filter.cpp

12tbuchman committed:
Remove unused covariance from cluster stats

12tbuchman committed:
Flip transforms in getMapCloud

12tbuchman committed:
Wrap target when resampling poses

12tbuchman committed:
Fix bug in enqueing occupancy map cells

12tbuchman committed:
Rename cached distances map distances_ for clarity

12tbuchman committed:
Add custom hash function to hash maps

12tbuchman committed:
Remove origin from octomap and use 0,0

12tbuchman committed:
Add comment to setOccDist about distances_end_

12tbuchman committed:
Make key_ a class vector for efficiency

12tbuchman committed:
Add bucket count to hashmaps at initialization

12tbuchman committed:
Add keysEqual function to hashmaps

12tbuchman committed:
Remove distances_end_ iterator

12tbuchman committed:
Remove distances_iterator_

12tbuchman committed:
Adding logging

12tbuchman committed:
Add distance value to OctoMapCellData

OctoMapCellData no longer needs to keep an instance of OctoMap
and can do a comparison directly with the distance value.

12tbuchman committed:
Remove logging comments

12tbuchman committed:
Replace priority queue with queue

OctoMapCellData no longer needs to know about the cell occ dist
and no longer needs to overload the comparison operator.

12tbuchman committed:
Fix bug in setting latest_tf_valid_

12tbuchman committed:
Improve hash map and related functions

Improving makeHash function to use prime coefficients and sum prime * key[i] values as hash.

Improving keysEqual to compare vectors using base implementation.

Using Eigen::Vector3i instead of std::Vector.

Increasing starting bucket count by 4x.

c-andy-martin reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 34 of 34 files at r1.
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-:-M5wF83DAsLCS1OW8Gly:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

12tbuchman committed:
Pass max occ dist to map directly from node

We were passing the max occ dist first to the sensor and having
the sensor pass it to the map. This is illogical as the sensor
doesn't care about the max occ dist.

12tbuchman committed:
Use individual int input parameters in isPoseValid

This makes it clear a pose is an 2d and not 3d.

Cropped cells are inclusive in range which is now reflected.

12tbuchman committed:
Call enqueue directly from iterateEmptyCells

We can skip the updateNode function and call enqueue directly with simpler parameters.

CachedDistanceMap now stores the distance with resolution factored in.

12tbuchman committed:
Remove unneeded updateCSpace in reconfigure node3d

12tbuchman committed:
Convert octomap LUT from hash table to vectors

The hash table is not very scalable because it takes too much memory.
It takes less memory  to store the distance itself in uint8 size (as
a ratio to the max distance). Additionally, to save memory we can only
store the values for a column if any z value of that column is anything
other than max occ dist. While this may not reduce the number of
voxels stored as much as the hashmap does, it probably reduces the majority.

Adding isVoxelValid to validate 3D input.

c-andy-martin reviewed (changes requested):
> Reviewed 15 of 15 files at r2.
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: all files reviewed, 3 unresolved discussions (waiting on @12tbuchman)

___
*[include/amcl/map/octomap.h, line 87 at r2](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M6Bm3n_4rDeIYt2AGH-:-M6Bm3n_4rDeIYt2AGH0:b3i7k6u) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/db913ec72ee6e03e409a3c260029fda65cf25e20/include/amcl/map/octomap.h#L87)):*
> ```objc
> 
>   std::shared_ptr<octomap::OcTree> octree_;
>   std::vector<uint32_t> pose_indices_;
> ```

I would prefer this to be called `xy_indices_` or something along those lines, as the entire pose is not an input, just the x and y positional parts (And related fields like `num_poses_` => `num_xy_indices_`
___
*[include/amcl/node/node_3d.h, line 130 at r2](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M6BlZhx0z6N1gtLymJ4:-M6BlZhy5hWG-Sk-yY5Q:b4xkadi) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/db913ec72ee6e03e409a3c260029fda65cf25e20/include/amcl/node/node_3d.h#L130)):*
> ```objc
>   double gompertz_input_scale_;
>   double gompertz_output_shift_;
>   double max_occ_dist_;
> ```

I'm not sure why we renamed here, we should only need the cspace for sensor models that require it. For instance, the beam model does not require cspace (as it raycasts through the map).
___
*[src/amcl/map/octomap.cpp, line 176 at r2](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M6Bmufe7daLbkMU__BH:-M6Bmufe7daLbkMU__BI:b-z4byq6) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/db913ec72ee6e03e409a3c260029fda65cf25e20/src/amcl/map/octomap.cpp#L176)):*
> ```cpp
>   distance_ratios_.clear();
>   distance_ratios_.resize(num_z_column_indices_, DTYPE_MAX);
>   distance_ratios_.shrink_to_fit();
> ```

I think it would be better here to take a stab at guessing how many you will need. and not shrink this at all, something like change this line to:
```
// Go ahead and reserve enough distance ratio space for a map that is only 1/16 occupied
distance_ratios_.reserve(num_z_column_indices_ * (num_poses_ / 16));
```


<!-- Sent from Reviewable.io -->

12tbuchman reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: all files reviewed, 3 unresolved discussions (waiting on @c-andy-martin)

___
*[include/amcl/map/octomap.h, line 87 at r2](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M6Bm3n_4rDeIYt2AGH-:-M6C4FBD514DrLUbFEJM:b-amlo2y) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/db913ec72ee6e03e409a3c260029fda65cf25e20/include/amcl/map/octomap.h#L87)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

I would prefer this to be called `xy_indices_` or something along those lines, as the entire pose is not an input, just the x and y positional parts (And related fields like `num_poses_` => `num_xy_indices_`
</blockquote></details>

pose is something consistent throughout badger_amcl relating to the xy indices (specifically not z). if i change that here id have to change it everywhere, but i think using pose is a better name than xy because our pose is not related to z at all. if we wanted in the future to make z a part of the pose, that would be when to change this.
___
*[include/amcl/node/node_3d.h, line 130 at r2](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M6BlZhx0z6N1gtLymJ4:-M6C4llGDrCfM44ujsjn:b-4l5ji9) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/db913ec72ee6e03e409a3c260029fda65cf25e20/include/amcl/node/node_3d.h#L130)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

I'm not sure why we renamed here, we should only need the cspace for sensor models that require it. For instance, the beam model does not require cspace (as it raycasts through the map).
</blockquote></details>

the beam model is a 2d implementation
___
*[src/amcl/map/octomap.cpp, line 176 at r2](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M6Bmufe7daLbkMU__BH:-M6C4vjz7zK6Xozpf7C_:b-z5ka4v) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/db913ec72ee6e03e409a3c260029fda65cf25e20/src/amcl/map/octomap.cpp#L176)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

I think it would be better here to take a stab at guessing how many you will need. and not shrink this at all, something like change this line to:
```
// Go ahead and reserve enough distance ratio space for a map that is only 1/16 occupied
distance_ratios_.reserve(num_z_column_indices_ * (num_poses_ / 16));
```
</blockquote></details>

good idea, ill add this.


<!-- Sent from Reviewable.io -->

c-andy-martin reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: all files reviewed, 2 unresolved discussions (waiting on @12tbuchman and @c-andy-martin)

___
*[include/amcl/node/node_3d.h, line 130 at r2](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M6BlZhx0z6N1gtLymJ4:-M6G6MVi4fvrpTNJ7Byr:bh08o0e) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/db913ec72ee6e03e409a3c260029fda65cf25e20/include/amcl/node/node_3d.h#L130)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

the beam model is a 2d implementation
</blockquote></details>

but we may want to add a beam model for 3D. Was the only motivation to simplify? Or is there a need to create cspace for other reasons beyond the sensor model? (thinking about it, a beam model would work much better in 3D than in our synthesized 2D scans)


<!-- Sent from Reviewable.io -->

12tbuchman committed:
Add obstacle cells to intermediate priority queue

This will mean iteration over empty cells happens in i, j, k order.

12tbuchman committed:
Delete octree after enqueing obstacle cells

12tbuchman committed:
Publish point cloud of cspace distances

12tbuchman committed:
Refactor loadPoseFromServer function in node.cpp

12tbuchman committed:
Fix bug calculating fsi in node 2d

12tbuchman committed:
Change name of saved pose file

The new file differs from the previous saved pose filename to avoid collisions.

12tbuchman committed:
Improve performance by globalizing vectors

Commit:
fix crash on shutdown due to destruction order

Provide destructors for the classes with tf::MessageFilters to be sure
they are destroyed prior to their associated subscribers.

12tbuchman committed:
Send initial pose internally

The initial pose does not need to be published at startup. It can be sent internally. No program should
rely on the initial pose being published at startup.

12tbuchman committed:
Set default data topics to universal names

12tbuchman committed:
In node.cpp waitForTransform, use scan timestamp

12tbuchman committed:
Change default save pose to server rate to 0.5hz

c-andy-martin reviewed (changes requested):
> Reviewed 8 of 12 files at r3, 2 of 4 files at r4, 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: all files reviewed, 15 unresolved discussions (waiting on @12tbuchman)

___
*[cfg/AMCL.cfg, line 125 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IOaM59BNnutai2wHC:-M7IOaM59BNnutai2wHD:bqu0uxe) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/cfg/AMCL.cfg#L125)):*
> ```INI
> 
> gen.add("save_pose", bool_t, 0, "If the node should save the pose to the param server and to a persistent file.", True)
> gen.add("saved_pose_filepath", str_t, 0, "Path of file to store saved poses.", "savedpose_amcl.yaml")
> ```

I slightly prefer "amcl_saved_pose.yaml" or perhaps even "bager_amcl_saved_pose.yaml".
___
*[src/amcl/map/octomap.cpp, line 176 at r2](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M6Bmufe7daLbkMU__BH:-M7IPQPKAUvO4cjb78mF:b-cc4u1t) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/db913ec72ee6e03e409a3c260029fda65cf25e20/src/amcl/map/octomap.cpp#L176)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

good idea, ill add this.
</blockquote></details>

IOW, do not call `shrink_to_fit` here and instead call `reserve` w/ a good guess. If you want to save memory, do the `shrink_to_fit` once you are completely done building the cspace. This way there are fewer allocations and copies as you go, which saves time
___
*[src/amcl/map/octomap.cpp, line 53 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IP6KiBKB1GiP8VFee:-M7IP6KiBKB1GiP8VFef:b-ep1dcu) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L53)):*
> ```cpp
>   octree_ = octree;
>   max_occ_dist_ = max_occ_dist;
>   max_occ_dist_ratio_ = max_occ_dist_ / UINT8_MAX;
> ```

replace `UINT8_MAX` with `std::numeric_limits<uint8_t>::max()`, and be sure `<limits>` is included
___
*[src/amcl/map/octomap.cpp, line 184 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IPJiK8Auywl8dAGfn:-M7IPJiK8Auywl8dAGfo:b-irkhgv) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L184)):*
> ```cpp
>   pose_indices_.shrink_to_fit();
>   distance_ratios_.clear();
>   distance_ratios_.resize(num_z_column_indices_, UINT8_MAX);
> ```

here too, everywhere use std::numeric_limits over the old C-style macros
___
*[src/amcl/map/octomap.cpp, line 196 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IZ-d-0CnWtp9BI-w4:-M7IZ-d-0CnWtp9BI-w5:bo6i3r3) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L196)):*
> ```cpp
>   iterateEmptyCells(q);
>   ROS_INFO("Done updating OctoMap CSpace");
>   publishDistances();
> ```

I think we should only do this conditional on some kind of debug setting. Could be a static private parameter that defaults to off. Alternatively, you could leave it always on and be sure to blacklist the topic so that blackbox doesn't record it.
___
*[src/amcl/map/octomap.cpp, line 208 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IPlzo6uWyIPmoMkLW:-M7IPlzo6uWyIPmoMkLX:b-rjzp5i) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L208)):*
> ```cpp
>   std::vector<int> map_coords(3);
> 
>   std::priority_queue<std::vector<int>> ordering_queue;
> ```

Did you try this w/ Eigen types instead of std::vector? It may be considerably faster (but the ordering queue may not be your limiting factor, however, IIRC, it took several seconds for walmart to add this x/y/z order-preserving step)
___
*[src/amcl/map/octomap.cpp, line 278 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IQ6TW9HS2q2ncnYQQ:-M7IQ6TW9HS2q2ncnYQR:b-mwce6e) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L278)):*
> ```cpp
>     count = std::max(count, q.size());
>   }
>   ROS_INFO("Max queue size: %lu", count);
> ```

I strongly prefer `ROS_INFO_STREAM` for most use cases. The only time I would use printf-style is when doing a lot of number formatting like `%8.2f` as its very verbose to do the formatting w/ C++ style iostreams. I'd prefer this here:
`ROS_INFO_STREAM("Max queue size: " << count);`
___
*[src/amcl/map/octomap.cpp, line 320 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IQbpNCqYgoym61BG-:-M7IQbpNCqYgoym61BG0:b-o03y1f) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L320)):*
> ```cpp
>     distance_ratios_.resize(distances_start_index + num_z_column_indices_, UINT8_MAX);
>   }
>   uint8_t distance_ratio = std::min(UINT8_MAX, static_cast<int>(std::floor(d / max_occ_dist_ * (UINT8_MAX + 1))));
> ```

so I wouldn't rely on UINT8_MAX (or the numeric limits version) being able to represent the `+1`, nor do I understand why you have a `+1` in the denom as a number between d and (d * 255/256) will all map to the max ratio of 255 . I think you should just remove the `+ 1`. Also, I would prefer clamping d first instead of the result, so something like this:
```
assert(d >= 0.0);
d = std::min(max_occ_dist_, d);
uint8_t distance_ratio = static_cast<uint8_t>(std::floor((d / max_occ_dist_) * std::numeric_limits<uint8_t>::max()));
```
You are guaranteed that when d == max_occ_dist_, the resulting double will be exactly 1.0, and also you are guaranteed that floor( 1.0 * 255 ) cast to an 8-bit unsigned will be 255. (Note this would not be true for a 32-bit integer and a 32-bit float. Either 32 or 64 bit floats have enough bits to exactly store the 8-bit number).
___
*[src/amcl/map/octomap.cpp, line 386 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IZKlM0Xj-FIRrWcke:-M7IZKlM0Xj-FIRrWckf:b-3489z3) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L386)):*
> ```cpp
>   distances_pub_ = nh_.advertise<PointCloud>("distances_cloud", 1, true);
>   distances_pub_.publish(cloud);
>   ROS_INFO("publishing cloud of size: %lu", cloud->points.size());
> ```

again, I like `ROS_INFO_STREAM` over printf-style for simple things like this
___
*[src/amcl/node/node.cpp, line 78 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IZRqL5oPDbU6qqjDM:-M7IZRqL5oPDbU6qqjDN:b-w9oes9) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node.cpp#L78)):*
> ```cpp
> 
>   private_nh_.param("save_pose", save_pose_, false);
>   const std::string default_filepath = "savedpose_amcl.yaml";
> ```

if you change the default filepath in the .cfg, do so here too
___
*[src/amcl/node/node.cpp, line 493 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IaUVwBCfMAb1giiH9:-M7IaUVwBCfMAb1giiHA:bkblb78) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node.cpp#L493)):*
> ```cpp
>   else
>   {
>     ROS_DEBUG_STREAM("Failed to load " << param_name << " from server.");
> ```

Seems like `ROS_WARN_STREAM` or at least info might be more appropriate
___
*[src/amcl/node/node_2d.cpp, line 74 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IaeWtE67nsF8pELNB:-M7IaeWtE67nsF8pELNC:b-bs0z8b) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_2d.cpp#L74)):*
> ```cpp
>   private_nh_.param("global_localization_planar_non_free_space_factor",
>                     global_localization_non_free_space_factor_, 1.0);
>   const std::string default_planar_scan_topic = "/scan";
> ```

this should not have the leading '/' but be in the private namespace, so it should just be `"scan"` (making it `"scan"` matches the old amcl)
___
*[src/amcl/node/node_2d.cpp, line 75 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IbHZTEDny4IXYF6Wx:-M7IbHZTEDny4IXYF6Wy:bbedzr1) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_2d.cpp#L75)):*
> ```cpp
>                     global_localization_non_free_space_factor_, 1.0);
>   const std::string default_planar_scan_topic = "/scan";
>   private_nh_.param("scan_topic", scan_topic_, default_planar_scan_topic);
> ```

BTW, I think we should remove these params for the topic names for `scan_topic` and `cloud_topic` and rely on remapping the topics to the appropriate place. Here is a discussion I found that provides some of the "why":
https://answers.ros.org/question/342777/is-it-correct-to-set-topic-names-from-parameters/
(And also, I am guilty of having added code that does this before I knew anything different).
We could also do the same for the odom_integrator_topic (which is something I added) and make it be "odom" and add remapping in the launch files to "/odom". The reason to make the topics in the private node namespace is because then you can easily remap it without affecting some other global topic name.
___
*[src/amcl/node/node_2d.cpp, line 327 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IeMz5DAVvMvFFEJuA:-M7IeMz5DAVvMvFFEJuB:blescv6) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_2d.cpp#L327)):*
> ```cpp
>     {
>       if (map_->getCellState(i, j) == MapCellState::CELL_FREE)
>       {
> ```

Why did we get rid of the non_free_space_radius_ check? Is it because we haven't setup the cspace yet? Not doing the check means randomFreeSpacePose won't work as well, as it won't honor the 'non_free_space_radius_' anymore. (Note this only matters when doing global localization as we have recovery disabled normally).
___
*[src/amcl/node/node_3d.cpp, line 77 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IfBEEBDprMySaUE-o:-M7IfBEEBDprMySaUE-p:b-yzod02) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_3d.cpp#L77)):*
> ```cpp
>   private_nh_.param("global_localization_scanner_non_free_space_factor",
>                     global_localization_non_free_space_factor_, 1.0);
>   const std::string default_point_cloud_scan_topic = "/cloud";
> ```

should be "cloud" so its in the private namespace, then remapped appropriately by the launch file(s)
___
*[src/amcl/node/node_3d.cpp, line 78 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IfFL7D684Y2EVVe0N:-M7IfFL7D684Y2EVVe0O:b-dy6ox0) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_3d.cpp#L78)):*
> ```cpp
>                     global_localization_non_free_space_factor_, 1.0);
>   const std::string default_point_cloud_scan_topic = "/cloud";
>   private_nh_.param("cloud_topic", scan_topic_, default_point_cloud_scan_topic);
> ```

see my comment on scan_topic, I'd like to remove these


<!-- Sent from Reviewable.io -->

12tbuchman committed:
Make publishCSpace a debug option

12tbuchman committed:
Reserve approximate size of cspace vector apriori

12tbuchman committed:
Fix bug in setOccDist calculating distance_ratio

12tbuchman committed:
Changing name of saved pose file.

12tbuchman committed:
Print an info message on loadPoseFromServer fail

12tbuchman committed:
Put scan and cloud topic default values in ns

12tbuchman committed:
Create free space indices after creating cspace

12tbuchman reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: all files reviewed, 15 unresolved discussions (waiting on @c-andy-martin)

___
*[cfg/AMCL.cfg, line 125 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IOaM59BNnutai2wHC:-M7IuMNTEa847k1O5fDW:bb71kmx) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/cfg/AMCL.cfg#L125)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

I slightly prefer "amcl_saved_pose.yaml" or perhaps even "bager_amcl_saved_pose.yaml".
</blockquote></details>

Okay
___
*[include/amcl/node/node_3d.h, line 130 at r2](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M6BlZhx0z6N1gtLymJ4:-M7Ip7_k2_PL4pv73t7P:b749xab) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/db913ec72ee6e03e409a3c260029fda65cf25e20/include/amcl/node/node_3d.h#L130)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

but we may want to add a beam model for 3D. Was the only motivation to simplify? Or is there a need to create cspace for other reasons beyond the sensor model? (thinking about it, a beam model would work much better in 3D than in our synthesized 2D scans)
</blockquote></details>

just to be clear, youre talking about renaming "sensor_likelihood_max_dist_" to "max_occ_dist_", right? I dont understand why that renaming would make the cspace be required. the reason i renamed it is because I separated creating the cspace from the sensor itself. theres no reason to backdoor the cspace through the sensor. It makes it more complicated to trace the creation of the cspace and in my opinion breaks the integrity of the node. the purpose of node is to manage the different parts, so the sensor should not be manipulating the map at all. we still keep a pointer to the map in the sensor to receive information from the map but the sensor is no longer manipulating the map. if in the future we dont want to create the cspace based on the model we can manage all of that in the node. we could even set up that infrastructure now in anticipation of a model that doesn't need a cspace.
___
*[src/amcl/map/octomap.cpp, line 176 at r2](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M6Bmufe7daLbkMU__BH:-M7Iq6r50RG9u2YsvZkI:b-tth29m) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/db913ec72ee6e03e409a3c260029fda65cf25e20/src/amcl/map/octomap.cpp#L176)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

IOW, do not call `shrink_to_fit` here and instead call `reserve` w/ a good guess. If you want to save memory, do the `shrink_to_fit` once you are completely done building the cspace. This way there are fewer allocations and copies as you go, which saves time
</blockquote></details>

sorry, i forgot to add this. its there now.
___
*[src/amcl/map/octomap.cpp, line 53 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IP6KiBKB1GiP8VFee:-M7Ipf7GBPkg0jMByYah:b-f9deix) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L53)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

replace `UINT8_MAX` with `std::numeric_limits<uint8_t>::max()`, and be sure `<limits>` is included
</blockquote></details>

<limits> is already included. why is it better to use std::numeric_limits though? seems unnecessarily complicated to use over a predefined constant.
___
*[src/amcl/map/octomap.cpp, line 184 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IPJiK8Auywl8dAGfn:-M7IpyU7-djtfTI66RoX:b-mmcvo8) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L184)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

here too, everywhere use std::numeric_limits over the old C-style macros
</blockquote></details>

Same question as before
___
*[src/amcl/map/octomap.cpp, line 196 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IZ-d-0CnWtp9BI-w4:-M7IxzUc6UgorP-aITui:b-896fix) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L196)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

I think we should only do this conditional on some kind of debug setting. Could be a static private parameter that defaults to off. Alternatively, you could leave it always on and be sure to blacklist the topic so that blackbox doesn't record it.
</blockquote></details>

Done.
___
*[src/amcl/map/octomap.cpp, line 208 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IPlzo6uWyIPmoMkLW:-M7JHLwZBL0tsV3IinUI:b-7zmgro) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L208)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

Did you try this w/ Eigen types instead of std::vector? It may be considerably faster (but the ordering queue may not be your limiting factor, however, IIRC, it took several seconds for walmart to add this x/y/z order-preserving step)
</blockquote></details>

good idea, ill give that a try.
___
*[src/amcl/map/octomap.cpp, line 278 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IQ6TW9HS2q2ncnYQQ:-M7IqGd92fnQE6YxLuh3:bcc0fua) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L278)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

I strongly prefer `ROS_INFO_STREAM` for most use cases. The only time I would use printf-style is when doing a lot of number formatting like `%8.2f` as its very verbose to do the formatting w/ C++ style iostreams. I'd prefer this here:
`ROS_INFO_STREAM("Max queue size: " << count);`
</blockquote></details>

When we talked before you said to use stream when printing a string and otherwise to not use stream... Im not really understanding the benefit of using a stream here. If I change this do you want me to go through and change every C style to C++ style throughout badger_amcl?
___
*[src/amcl/map/octomap.cpp, line 320 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IQbpNCqYgoym61BG-:-M7J3Z4339VpKW2EH8Su:bu1vldx) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L320)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

so I wouldn't rely on UINT8_MAX (or the numeric limits version) being able to represent the `+1`, nor do I understand why you have a `+1` in the denom as a number between d and (d * 255/256) will all map to the max ratio of 255 . I think you should just remove the `+ 1`. Also, I would prefer clamping d first instead of the result, so something like this:
```
assert(d >= 0.0);
d = std::min(max_occ_dist_, d);
uint8_t distance_ratio = static_cast<uint8_t>(std::floor((d / max_occ_dist_) * std::numeric_limits<uint8_t>::max()));
```
You are guaranteed that when d == max_occ_dist_, the resulting double will be exactly 1.0, and also you are guaranteed that floor( 1.0 * 255 ) cast to an 8-bit unsigned will be 255. (Note this would not be true for a 32-bit integer and a 32-bit float. Either 32 or 64 bit floats have enough bits to exactly store the 8-bit number).
</blockquote></details>

youre right, UINT8_MAX + 1 is not the right way to do this.

the reason I was multiplying by max+1 instead of max is because this makes a pseudo rounding instead of floor. But I just changed std::floor to std::round which achieves what I want in a simpler way and without going beyond the numerical limits of our integer.

As a protected function setOccDist is called internally, so we don't have to assert that its value is >=0. Based on implementation its value could be greater than the max occ dist, and I agree its better to constrain it before calculating the ratio.
___
*[src/amcl/map/octomap.cpp, line 386 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IZKlM0Xj-FIRrWcke:-M7IqtKu1EKQimd2iyMv:b-w2ufq0) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L386)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

again, I like `ROS_INFO_STREAM` over printf-style for simple things like this
</blockquote></details>

see previous comment
___
*[src/amcl/node/node.cpp, line 78 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IZRqL5oPDbU6qqjDM:-M7IrE0o88fp-cnIMC9J:bb7f2mf) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node.cpp#L78)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

if you change the default filepath in the .cfg, do so here too
</blockquote></details>

done
___
*[src/amcl/node/node.cpp, line 493 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IaUVwBCfMAb1giiH9:-M7IrGN0-Kx2nLl9j_UB:bdptl9t) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node.cpp#L493)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

Seems like `ROS_WARN_STREAM` or at least info might be more appropriate
</blockquote></details>

This isnt a problem, we have the file as backup. Any time the robot reboots we won't be able to load the params from the param server. I made it a debug because i dont want any false alarms. I could make it an Info so that we see it even when its expected, but I wouldn't want to make it a WARN because most of the time that its printed its not a problem.
___
*[src/amcl/node/node_2d.cpp, line 74 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IaeWtE67nsF8pELNB:-M7Irlv9BeCOoc6JREpk:b-896fix) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_2d.cpp#L74)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

this should not have the leading '/' but be in the private namespace, so it should just be `"scan"` (making it `"scan"` matches the old amcl)
</blockquote></details>

Done.
___
*[src/amcl/node/node_2d.cpp, line 75 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IbHZTEDny4IXYF6Wx:-M7IteQ-3sWfFtyTr6Om:bh4fjns) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_2d.cpp#L75)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

BTW, I think we should remove these params for the topic names for `scan_topic` and `cloud_topic` and rely on remapping the topics to the appropriate place. Here is a discussion I found that provides some of the "why":
https://answers.ros.org/question/342777/is-it-correct-to-set-topic-names-from-parameters/
(And also, I am guilty of having added code that does this before I knew anything different).
We could also do the same for the odom_integrator_topic (which is something I added) and make it be "odom" and add remapping in the launch files to "/odom". The reason to make the topics in the private node namespace is because then you can easily remap it without affecting some other global topic name.
</blockquote></details>

Im not sure that discussion provided a solid answer to remapping vs parameters but rather provided a number of various opinions. If I understand correctly though, you want to remove the parameters and force remapping to avoid a clash between remapping and parameter names? My argument for keeping parameters in would be that the old amcl has topic parameters.
___
*[src/amcl/node/node_2d.cpp, line 327 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IeMz5DAVvMvFFEJuA:-M7JAlpzDPQqTyRJQJft:b-s6w3p1) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_2d.cpp#L327)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

Why did we get rid of the non_free_space_radius_ check? Is it because we haven't setup the cspace yet? Not doing the check means randomFreeSpacePose won't work as well, as it won't honor the 'non_free_space_radius_' anymore. (Note this only matters when doing global localization as we have recovery disabled normally).
</blockquote></details>

yeah, we dont have a cspace created yet so we cant call getOccDist. however, theres a disclaimer about that in the comments at the top of this function. The better solution is to call updateFreeSpaceIndices after creating the cspace.
___
*[src/amcl/node/node_3d.cpp, line 77 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IfBEEBDprMySaUE-o:-M7JFsca98JZ73fn7SpF:b-896fix) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_3d.cpp#L77)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

should be "cloud" so its in the private namespace, then remapped appropriately by the launch file(s)
</blockquote></details>

Done.
___
*[src/amcl/node/node_3d.cpp, line 78 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IfFL7D684Y2EVVe0N:-M7JFtbHBnEPwSZVkVqW:bhd9rm7) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_3d.cpp#L78)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

see my comment on scan_topic, I'd like to remove these
</blockquote></details>

see previous discussion.


<!-- Sent from Reviewable.io -->

12tbuchman committed:
Make struct EV3 for octomap cspace priority queue

12tbuchman committed:
Replace UINT8_MAX constant with numeric limit

12tbuchman reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: 28 of 35 files reviewed, 15 unresolved discussions (waiting on @c-andy-martin)

___
*[src/amcl/map/octomap.cpp, line 208 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IPlzo6uWyIPmoMkLW:-M7K7W80-4UnyZEE4JQq:b-896fix) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L208)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

good idea, ill give that a try.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

c-andy-martin commented:
___
*[src/amcl/node/node.cpp, line 493 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IaUVwBCfMAb1giiH9:-M7Nh4SH1gFXvKJDTIMy:b32w0x) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node.cpp#L493)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

This isnt a problem, we have the file as backup. Any time the robot reboots we won't be able to load the params from the param server. I made it a debug because i dont want any false alarms. I could make it an Info so that we see it even when its expected, but I wouldn't want to make it a WARN because most of the time that its printed its not a problem.
</blockquote></details>

OK


<!-- Sent from Reviewable.io -->

c-andy-martin reviewed (changes requested):
> Reviewed 4 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: 32 of 35 files reviewed, 9 unresolved discussions (waiting on @12tbuchman and @c-andy-martin)

___
*[include/amcl/map/octomap.h, line 49 at r6](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7NbM4--ocMvHW37ZF_:-M7NbM40B2cVjQCYbPvx:be9kw23) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/e18deb863b0b8d2363caa17c7cad81ea76b982df/include/amcl/map/octomap.h#L49)):*
> ```objc
> };
> 
> struct EV3 {
> ```

I'm not a huge fan of `EV3` as a type name as it is very dense, and doesn't communicate the underlying data-type of integer either. How about `Index3D`?
___
*[src/amcl/map/octomap.cpp, line 53 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IP6KiBKB1GiP8VFee:-M7NcwbQDH7nYfzZgX7r:bm6zs38) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L53)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

<limits> is already included. why is it better to use std::numeric_limits though? seems unnecessarily complicated to use over a predefined constant.
</blockquote></details>

It allows one to more easily change the data type. You can just search for `uint8_t` and replace it with `AwesomeDataType` and not break the code. Also, if you wanted to make the data type a template parameter in the future, you just drop in the template type to the limits. The old C-style macros are not templatized so you would have to know what the type was to get the appropriate limit.
___
*[src/amcl/map/octomap.cpp, line 278 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IQ6TW9HS2q2ncnYQQ:-M7NdVXaD9Z3SdGbi06b:b4da1uy) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L278)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

When we talked before you said to use stream when printing a string and otherwise to not use stream... Im not really understanding the benefit of using a stream here. If I change this do you want me to go through and change every C style to C++ style throughout badger_amcl?
</blockquote></details>

What I meant was, you can use ROS_INFO when there are no arguments, because there is no benefit to ROS_INFO_STREAM in that case. In general, _STREAM is always preferred when using arguments. There are 2 very good reasons for this. The first is that if you don't have warnings as errors when compiling and you mismatch printf style format specifiers with the wrong type of variable you can cause unpredictable behavior (like passing a float to a "%s"). I have had to track down several annoying SIGSEGVs over the years due to this exact issue where someone changed a debug print and didn't alter the format specifier list correctly. The second is that you have to know something about the type to correctly write the format specifier, while `<<` handles this automatically for you in the stream case, making it easier to templatize and maintain code as data types change. Imagine that some day there are 128 bit integers and that your have to modify format specifiers to match the change to the underlying type. This sort of thing happened when moving to 64-bit and was very painful for lots of C code that doesn't have the option.
___
*[src/amcl/map/octomap.cpp, line 320 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IQbpNCqYgoym61BG-:-M7NfT2t3MdRs0L-CSX5:b1i9f48) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L320)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

youre right, UINT8_MAX + 1 is not the right way to do this.

the reason I was multiplying by max+1 instead of max is because this makes a pseudo rounding instead of floor. But I just changed std::floor to std::round which achieves what I want in a simpler way and without going beyond the numerical limits of our integer.

As a protected function setOccDist is called internally, so we don't have to assert that its value is >=0. Based on implementation its value could be greater than the max occ dist, and I agree its better to constrain it before calculating the ratio.
</blockquote></details>

I use assert internally whenever doing so would prevent undefined behavior or behavior that would be hard to test, or to verify something that should be logically impossible just to be sure the logic stays correct (or was correct to start with). It is very easy to change the internals of a class later when doing maintenance or adding new features and forgetting the contract of the function being called. Tracing down undefined behavior is a lot harder than compiling the assertions on and failing an assert(). The way we compile our releases we compile assertions off, so adding assert() does not change the runtime behavior in a release. In this case, the behavior is not undefined, but is certainly undesirable and may be hard to catch or test, because passing a negative value (or a NAN which the assert would also catch) would result in odd behavior of the localizer.
___
*[src/amcl/node/node.cpp, line 427 at r6](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7NgwZb7VMliKox39iB:-M7NgwZb7VMliKox39iC:b-4x715k) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/e18deb863b0b8d2363caa17c7cad81ea76b982df/src/amcl/node/node.cpp#L427)):*
> ```cpp
>   if (loadPoseFromServer())
>   {
>     ROS_DEBUG("Successfully loaded pose from server.");
> ```

I think this should be ROS_INFO so someone checking the log output can see which case we hit. This might help debug any problems when the initial pose is not loaded properly.


<!-- Sent from Reviewable.io -->

c-andy-martin commented:
___
*[src/amcl/node/node_2d.cpp, line 327 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IeMz5DAVvMvFFEJuA:-M7Nhw0S96k4YwEF_XAW:b-n79f9s) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_2d.cpp#L327)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

yeah, we dont have a cspace created yet so we cant call getOccDist. however, theres a disclaimer about that in the comments at the top of this function. The better solution is to call updateFreeSpaceIndices after creating the cspace.
</blockquote></details>

Ah, excellent. Thanks!


<!-- Sent from Reviewable.io -->

c-andy-martin commented:
___
*[src/amcl/node/node_3d.cpp, line 78 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IfFL7D684Y2EVVe0N:-M7Ni5EB0Be0jYBjn-ct:b200h1k) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_3d.cpp#L78)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

see previous discussion.
</blockquote></details>

See my reply


<!-- Sent from Reviewable.io -->

c-andy-martin commented:
___
*[src/amcl/map/octomap.cpp, line 386 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IZKlM0Xj-FIRrWcke:-M7NiC70C1b6xhlBL1oI:b200h1k) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L386)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

see previous comment
</blockquote></details>

See my reply


<!-- Sent from Reviewable.io -->

c-andy-martin commented:
___
*[src/amcl/node/node_2d.cpp, line 75 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IbHZTEDny4IXYF6Wx:-M7NhTaRE4vcJy1eD744:b-38a1g2) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_2d.cpp#L75)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

Im not sure that discussion provided a solid answer to remapping vs parameters but rather provided a number of various opinions. If I understand correctly though, you want to remove the parameters and force remapping to avoid a clash between remapping and parameter names? My argument for keeping parameters in would be that the old amcl has topic parameters.
</blockquote></details>

The only topic parameter in the old amcl is one I added for the odom topic, which if I wrote it today I would not have done, unless I'm missing something. I would prefer to remove them all from badger_amcl and rely on remapping, yes (with the stipulation that all the topic names are in the private node namespace)


<!-- Sent from Reviewable.io -->

c-andy-martin reviewed (commented):
> Reviewed 3 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: all files reviewed, 6 unresolved discussions (waiting on @12tbuchman)



<!-- Sent from Reviewable.io -->

12tbuchman committed:
Centralize pose loaded info statements to loadPose()

12tbuchman committed:
Improve name of struct in octomap to Index3

12tbuchman committed:
Clean up ros console messages

12tbuchman committed:
Assert that d is >= 0 in octomap setOccDist

12tbuchman committed:
Remove topic name parameters

12tbuchman reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: all files reviewed, 6 unresolved discussions (waiting on @c-andy-martin)

___
*[include/amcl/map/octomap.h, line 49 at r6](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7NbM4--ocMvHW37ZF_:-M7NvBNq4WpJDeHTsQbl:by4y8e1) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/e18deb863b0b8d2363caa17c7cad81ea76b982df/include/amcl/map/octomap.h#L49)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

I'm not a huge fan of `EV3` as a type name as it is very dense, and doesn't communicate the underlying data-type of integer either. How about `Index3D`?
</blockquote></details>

I changed it to Index3
___
*[src/amcl/map/octomap.cpp, line 53 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IP6KiBKB1GiP8VFee:-M7NnLSU21DxSNKZRBvq:be8jdlk) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L53)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

It allows one to more easily change the data type. You can just search for `uint8_t` and replace it with `AwesomeDataType` and not break the code. Also, if you wanted to make the data type a template parameter in the future, you just drop in the template type to the limits. The old C-style macros are not templatized so you would have to know what the type was to get the appropriate limit.
</blockquote></details>

yeah I played with making it a template when I originally implemented this but decided uint8 was probably good in every situation. good to know that I can more easily template with this numeric limits though. thanks.
___
*[src/amcl/map/octomap.cpp, line 278 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IQ6TW9HS2q2ncnYQQ:-M7NyeIg4C9qVw_NiDUV:b-896fix) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L278)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

What I meant was, you can use ROS_INFO when there are no arguments, because there is no benefit to ROS_INFO_STREAM in that case. In general, _STREAM is always preferred when using arguments. There are 2 very good reasons for this. The first is that if you don't have warnings as errors when compiling and you mismatch printf style format specifiers with the wrong type of variable you can cause unpredictable behavior (like passing a float to a "%s"). I have had to track down several annoying SIGSEGVs over the years due to this exact issue where someone changed a debug print and didn't alter the format specifier list correctly. The second is that you have to know something about the type to correctly write the format specifier, while `<<` handles this automatically for you in the stream case, making it easier to templatize and maintain code as data types change. Imagine that some day there are 128 bit integers and that your have to modify format specifiers to match the change to the underlying type. This sort of thing happened when moving to 64-bit and was very painful for lots of C code that doesn't have the option.
</blockquote></details>

Done.
___
*[src/amcl/map/octomap.cpp, line 386 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IZKlM0Xj-FIRrWcke:-M7NzpB047THM9Tot24h:b-896fix) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/map/octomap.cpp#L386)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

See my reply
</blockquote></details>

Done.
___
*[src/amcl/node/node.cpp, line 493 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IaUVwBCfMAb1giiH9:-M7O-8av98_6OqcAYqS_:bb1j542) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node.cpp#L493)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

OK
</blockquote></details>

Im going to leave this as a debug actually. We already print that we failed to load the params from the server in general. Which parameter failed is really only important when debugging.
___
*[src/amcl/node/node.cpp, line 427 at r6](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7NgwZb7VMliKox39iB:-M7No6tO89FPIT_UMoo2:bxtzxpy) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/e18deb863b0b8d2363caa17c7cad81ea76b982df/src/amcl/node/node.cpp#L427)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

I think this should be ROS_INFO so someone checking the log output can see which case we hit. This might help debug any problems when the initial pose is not loaded properly.
</blockquote></details>

I currently have success messages in loadPoseFromServer. it might make more sense to centralize all of these comments to this function though.
___
*[src/amcl/node/node_3d.cpp, line 78 at r5](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-M7IfFL7D684Y2EVVe0N:-M7O-efAEnB0U3SXw7En:b-896fix) ([raw file](https://github.com/badgertechnologies/badger_amcl/blob/39003c4cef32a03c55e4fdd3ca1b7581c5e854a1/src/amcl/node/node_3d.cpp#L78)):*
<details><summary><i>Previously, c-andy-martin (C. Andy Martin) wrote…</i></summary><blockquote>

See my reply
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

c-andy-martin reviewed (approved):
> Reviewed 12 of 12 files at r7.
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

c-andy-martin reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3#-:-M7OLBoL2nSd5lw8BlIJ:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

12tbuchman committed:
Remove unneeded node handler from main.cpp

c-andy-martin reviewed (approved):
> Reviewed 3 of 3 files at r8.
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

12tbuchman committed:
Remove upstream ros tests

c-andy-martin reviewed (approved):
> Reviewed 16 of 16 files at r9.
> *[Reviewable](https://reviewable.io/reviews/badgertechnologies/badger_amcl/3)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->
